In [17]:
import pandas as pd
import pandasql as ps
import warnings
warnings.filterwarnings("ignore")
from datetime import datetime


tabela = pd.read_csv("C://Users/Cliente/Desktop/estágio/Case Estagiário/Questão 1 - Base.txt", sep="	")
display(tabela)


ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
0          402467    2019-01-01        FX0     210.00
1          403001    2019-01-01        FX1    1097.18
2          403499    2019-01-01        FX0    1209.88
3          403670    2019-01-01        FX0     667.40
4          404183    2019-01-01        FX1    1239.52
...           ...           ...        ...        ...
2728244   1624841    2019-08-08        FX1    1467.05
2728245   1624848    2019-08-08        FX0    1189.15
2728246   1624871    2019-08-08        FX1     427.72
2728247   1624878    2019-08-08        FX0     281.14
2728248   1624884    2019-08-08        FX0     313.00

[2728249 rows x 4 columns]

In [4]:
query = """select * from tabela """
ps.sqldf(query, locals())

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
0          402467    2019-01-01        FX0     210.00
1          403001    2019-01-01        FX1    1097.18
2          403499    2019-01-01        FX0    1209.88
3          403670    2019-01-01        FX0     667.40
4          404183    2019-01-01        FX1    1239.52
...           ...           ...        ...        ...
2728244   1624841    2019-08-08        FX1    1467.05
2728245   1624848    2019-08-08        FX0    1189.15
2728246   1624871    2019-08-08        FX1     427.72
2728247   1624878    2019-08-08        FX0     281.14
2728248   1624884    2019-08-08        FX0     313.00

[2728249 rows x 4 columns]

In [5]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela
where DS_ROLAGEM ='FX1' """
ps.sqldf(query, locals())

DT_VENCIMENTO DS_ROLAGEM
0         2019-01-01        FX1
1         2019-01-01        FX1
2         2019-01-01        FX1
3         2019-01-01        FX1
4         2019-01-01        FX1
...              ...        ...
246448    2019-08-08        FX1
246449    2019-08-08        FX1
246450    2019-08-08        FX1
246451    2019-08-08        FX1
246452    2019-08-08        FX1

[246453 rows x 2 columns]

In [6]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-01-%' """
mes_janeiro= ps.sqldf(query, locals())

display(mes_janeiro)
print(mes_janeiro["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-01-01        FX0
1         2019-01-01        FX1
2         2019-01-01        FX0
3         2019-01-01        FX0
4         2019-01-01        FX1
...              ...        ...
313378    2019-01-08        FX0
313379    2019-01-08        FX0
313380    2019-01-08        FX0
313381    2019-01-08        FX0
313382    2019-01-08        FX0

[313383 rows x 2 columns]

FX0    92.7%
FX1     7.3%
Name: DS_ROLAGEM, dtype: object


In [7]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-02-%' """
mes_fevereiro = ps.sqldf(query, locals())

display(mes_fevereiro)
print(mes_fevereiro["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-02-01        FX0
1         2019-02-01        FX0
2         2019-02-01        FX0
3         2019-02-01        FX0
4         2019-02-01        FX0
...              ...        ...
313844    2019-02-28        FX0
313845    2019-02-28        FX0
313846    2019-02-28        FX0
313847    2019-02-28        FX0
313848    2019-02-28        FX0

[313849 rows x 2 columns]

FX0    91.1%
FX1     8.9%
Name: DS_ROLAGEM, dtype: object


In [8]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-03-%' """
mes_marco = ps.sqldf(query, locals())

display(mes_marco)
print(mes_marco["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-03-01        FX0
1         2019-03-01        FX1
2         2019-03-01        FX0
3         2019-03-01        FX0
4         2019-03-01        FX0
...              ...        ...
307313    2019-03-28        FX0
307314    2019-03-28        FX0
307315    2019-03-28        FX0
307316    2019-03-28        FX1
307317    2019-03-28        FX0

[307318 rows x 2 columns]

FX0    89.1%
FX1    10.9%
Name: DS_ROLAGEM, dtype: object


In [9]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-04-%' """
mes_abril= ps.sqldf(query, locals())

display(mes_abril)
print(mes_abril["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-04-01        FX0
1         2019-04-01        FX0
2         2019-04-01        FX0
3         2019-04-01        FX0
4         2019-04-01        FX0
...              ...        ...
302325    2019-04-28        FX0
302326    2019-04-28        FX0
302327    2019-04-28        FX0
302328    2019-04-28        FX0
302329    2019-04-28        FX0

[302330 rows x 2 columns]

FX0    91.6%
FX1     8.4%
Name: DS_ROLAGEM, dtype: object


In [10]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-05-%' """
mes_maio= ps.sqldf(query, locals())

display(mes_maio)
print(mes_maio["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-05-01        FX0
1         2019-05-01        FX1
2         2019-05-01        FX0
3         2019-05-01        FX0
4         2019-05-01        FX0
...              ...        ...
301860    2019-05-28        FX0
301861    2019-05-28        FX0
301862    2019-05-28        FX0
301863    2019-05-28        FX0
301864    2019-05-28        FX0

[301865 rows x 2 columns]

FX0    90.0%
FX1    10.0%
Name: DS_ROLAGEM, dtype: object


In [11]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-06-%' """
mes_junho= ps.sqldf(query, locals())

display(mes_junho)
print(mes_junho["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-06-01        FX0
1         2019-06-01        FX0
2         2019-06-01        FX0
3         2019-06-01        FX0
4         2019-06-01        FX0
...              ...        ...
304361    2019-06-28        FX0
304362    2019-06-28        FX0
304363    2019-06-28        FX0
304364    2019-06-28        FX0
304365    2019-06-28        FX0

[304366 rows x 2 columns]

FX0    91.5%
FX1     8.5%
Name: DS_ROLAGEM, dtype: object


In [12]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-07-%' """
mes_julho= ps.sqldf(query, locals())

display(mes_julho)
print(mes_julho["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-07-01        FX1
1         2019-07-01        FX0
2         2019-07-01        FX0
3         2019-07-01        FX0
4         2019-07-01        FX0
...              ...        ...
310889    2019-07-28        FX0
310890    2019-07-28        FX0
310891    2019-07-28        FX1
310892    2019-07-28        FX0
310893    2019-07-28        FX0

[310894 rows x 2 columns]

FX0    90.4%
FX1     9.6%
Name: DS_ROLAGEM, dtype: object


In [13]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-08-%' """
mes_agosto= ps.sqldf(query, locals())

display(mes_agosto)
print(mes_agosto["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-08-08        FX0
1         2019-08-08        FX0
2         2019-08-08        FX0
3         2019-08-08        FX0
4         2019-08-08        FX0
...              ...        ...
317062    2019-08-08        FX1
317063    2019-08-08        FX0
317064    2019-08-08        FX1
317065    2019-08-08        FX0
317066    2019-08-08        FX0

[317067 rows x 2 columns]

FX0    90.9%
FX1     9.1%
Name: DS_ROLAGEM, dtype: object


In [14]:
query = """ select DT_VENCIMENTO, DS_ROLAGEM from tabela 
where DT_VENCIMENTO like '%-09-%' """
mes_setembro= ps.sqldf(query, locals())

display(mes_setembro)
print(mes_setembro["DS_ROLAGEM"].value_counts(normalize=True).map("{:.1%}".format))

DT_VENCIMENTO DS_ROLAGEM
0         2019-09-01        FX0
1         2019-09-01        FX0
2         2019-09-01        FX0
3         2019-09-01        FX0
4         2019-09-01        FX0
...              ...        ...
257172    2019-09-23        FX0
257173    2019-09-23        FX0
257174    2019-09-23        FX0
257175    2019-09-23        FX0
257176    2019-09-23        FX0

[257177 rows x 2 columns]

FX0    91.5%
FX1     8.5%
Name: DS_ROLAGEM, dtype: object


# Temos que o percentual de faturas emitidas por mês no qual os clientes não pagaram a fatura anterior é de:

## Janeiro= 7.3%
## Fevereiro= 8.9%
## Março= 10.9%
## Abril= 8.4%
## Maio= 10%
## Junho= 8.5%
## Julho= 9.6%
## Agosto= 9.1%
## Setembro= 8.5%

### Fazendo uma breve análise temos que o maior percentual dos clientes que não pagaram a fatura é em março, com isso poderia fazer-se uma análise nesse mês para saber a causa do aumento no mês de março. 

In [15]:
query = """select * from tabela """
ps.sqldf(query, locals())

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
0          402467    2019-01-01        FX0     210.00
1          403001    2019-01-01        FX1    1097.18
2          403499    2019-01-01        FX0    1209.88
3          403670    2019-01-01        FX0     667.40
4          404183    2019-01-01        FX1    1239.52
...           ...           ...        ...        ...
2728244   1624841    2019-08-08        FX1    1467.05
2728245   1624848    2019-08-08        FX0    1189.15
2728246   1624871    2019-08-08        FX1     427.72
2728247   1624878    2019-08-08        FX0     281.14
2728248   1624884    2019-08-08        FX0     313.00

[2728249 rows x 4 columns]

In [18]:
display(tabela)

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
0          402467    2019-01-01        FX0     210.00
1          403001    2019-01-01        FX1    1097.18
2          403499    2019-01-01        FX0    1209.88
3          403670    2019-01-01        FX0     667.40
4          404183    2019-01-01        FX1    1239.52
...           ...           ...        ...        ...
2728244   1624841    2019-08-08        FX1    1467.05
2728245   1624848    2019-08-08        FX0    1189.15
2728246   1624871    2019-08-08        FX1     427.72
2728247   1624878    2019-08-08        FX0     281.14
2728248   1624884    2019-08-08        FX0     313.00

[2728249 rows x 4 columns]

In [19]:
tab=tabela.loc[(tabela['DT_VENCIMENTO']>='2019-09-01') & (tabela['DT_VENCIMENTO']<='2019-09-30')]

In [20]:
tab.loc[tab['ID_CONTA'] == 1598850]

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
998014    1598850    2019-09-01        FX0     533.54
1472790   1598850    2019-09-23        FX0     596.95

In [21]:
tab.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 257177 entries, 967588 to 1487074
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   ID_CONTA       257177 non-null  int64  
 1   DT_VENCIMENTO  257177 non-null  object 
 2   DS_ROLAGEM     257177 non-null  object 
 3   VL_FATURA      257177 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 9.8+ MB


In [23]:
tab.head()

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
967588     99416    2019-09-01        FX0    2860.79
967589     52706    2019-09-01        FX0     180.00
967590   1024221    2019-09-01        FX0     813.28
967591   1024238    2019-09-01        FX0     808.15
967592   1024293    2019-09-01        FX0    2068.17

In [24]:
tab1 = tabela.loc[(tabela['DT_VENCIMENTO']>='2019-03-01') & (tabela['DT_VENCIMENTO'] <= '2019-08-31')]

In [25]:
ndf = tab1[tab1['ID_CONTA'].isin(tab['ID_CONTA'])]
ndf.loc[ndf['ID_CONTA']==205737]

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
766878     205737    2019-03-01        FX0     492.77
1393013    205737    2019-04-01        FX0     535.17
1746606    205737    2019-05-01        FX0     482.94
2035494    205737    2019-06-01        FX0     409.52
2347449    205737    2019-07-01        FX0     468.34
2660994    205737    2019-08-01        FX0     288.00

In [26]:
ndf[ndf['ID_CONTA']==755517]

ID_CONTA DT_VENCIMENTO DS_ROLAGEM  VL_FATURA
771936     755517    2019-08-20        FX1    1205.28
1171292    755517    2019-03-20        FX1     665.22
1659771    755517    2019-04-20        FX1     687.35
1949024    755517    2019-05-20        FX1     737.58
2265719    755517    2019-06-20        FX0     517.42
2565209    755517    2019-07-20        FX1     982.78

In [27]:
ndf1 = ndf.loc[ndf['DS_ROLAGEM']=='FX1']

In [28]:
ndf2 = ndf1.groupby('ID_CONTA',as_index=False)['DS_ROLAGEM'].count()

In [29]:
ndf2[ndf2['ID_CONTA']==1624908]

ID_CONTA  DS_ROLAGEM
42276   1624908           2

In [31]:
df0 = tab1[tab1['ID_CONTA'].isin(tab['ID_CONTA'])]
df0.drop(['DT_VENCIMENTO','DS_ROLAGEM'],axis=1,inplace=True)
df_teste = pd.merge(df0,tab[['ID_CONTA','DT_VENCIMENTO','DS_ROLAGEM']],on='ID_CONTA')
df_teste = pd.merge(df_teste,ndf2,on='ID_CONTA')

In [32]:
df0.head()

ID_CONTA  VL_FATURA
464874   1624902     226.62
464875   1624908     921.09
464876   1624916    1456.75
464877   1624951      70.11
464878   1624975      49.05

In [33]:
a = df_teste.groupby('ID_CONTA',as_index=False).count()
a.drop(a.columns.difference(['ID_CONTA',"DT_VENCIMENTO"]),axis=1,inplace=True)
b = df_teste.groupby('ID_CONTA',as_index=False)['VL_FATURA'].mean()
df1 = pd.merge(a,b,on='ID_CONTA')
df1.rename(columns={'DT_VENCIMENTO': 'QTD_FATURAS_ULT_6M', 'VL_FATURA': 'VL_MEDIO_FATURA'}, inplace=True)

In [34]:
df2 = pd.merge(df_teste,df1,on='ID_CONTA')

In [35]:
df2.head()

ID_CONTA  VL_FATURA DT_VENCIMENTO DS_ROLAGEM_x  DS_ROLAGEM_y  \
0   1624908     921.09    2019-09-08          FX1             2   
1   1624908     881.45    2019-09-08          FX1             2   
2   1624908    1474.28    2019-09-08          FX1             2   
3   1624908    2288.68    2019-09-08          FX1             2   
4   1624908     478.23    2019-09-08          FX1             2   

   QTD_FATURAS_ULT_6M  VL_MEDIO_FATURA  
0                   6      1188.201667  
1                   6      1188.201667  
2                   6      1188.201667  
3                   6      1188.201667  
4                   6      1188.201667

In [36]:
df2['DT_VENCIMENTO'] = pd.to_datetime(df2['DT_VENCIMENTO'])
df2.sort_values(by='DT_VENCIMENTO',ascending=True,inplace=True);df2.head()

ID_CONTA  VL_FATURA DT_VENCIMENTO DS_ROLAGEM_x  DS_ROLAGEM_y  \
166533   1505179     364.58    2019-09-01          FX0             1   
158415   1389400    1114.83    2019-09-01          FX0             2   
158416   1389400    1033.57    2019-09-01          FX0             2   
158417   1389400    1100.55    2019-09-01          FX0             2   
158418   1389400     333.51    2019-09-01          FX0             2   

        QTD_FATURAS_ULT_6M  VL_MEDIO_FATURA  
166533                   6       439.410000  
158415                   6       824.546667  
158416                   6       824.546667  
158417                   6       824.546667  
158418                   6       824.546667

In [37]:
df2.drop_duplicates(subset=['ID_CONTA'],inplace=True,ignore_index=True)

In [38]:
lista_cols = ['ID_CONTA','DS_ROLAGEM_x','DT_VENCIMENTO','QTD_FATURAS_ULT_6M','VL_MEDIO_FATURA','DS_ROLAGEM_y']

for cols in df2.columns:
  if cols in lista_cols:
    None
  else: 
    df2.drop([cols],axis=1,inplace=True) 

df2.rename(columns={'DS_ROLAGEM_x': 'DS_ROLAGEM', 'DT_VENCIMENTO': 'DT_VENCIMENTO',
                   'DS_ROLAGEM_y': 'QTD_FATURAS_ULT_6M_FX1'}, inplace=True)
df2 = df2.reindex(columns=['ID_CONTA','DS_ROLAGEM','DT_VENCIMENTO','QTD_FATURAS_ULT_6M','VL_MEDIO_FATURA','QTD_FATURAS_ULT_6M_FX1'])

In [39]:
display(df2)

ID_CONTA DS_ROLAGEM DT_VENCIMENTO  QTD_FATURAS_ULT_6M  VL_MEDIO_FATURA  \
0       1505179        FX0    2019-09-01                   6       439.410000   
1       1389400        FX0    2019-09-01                   6       824.546667   
2       1389402        FX1    2019-09-01                   6       510.290000   
3       1389539        FX0    2019-09-01                   6       534.598333   
4       1389990        FX0    2019-09-01                   6       315.980000   
...         ...        ...           ...                 ...              ...   
47432   1511980        FX1    2019-09-23                   5       233.082000   
47433   1512071        FX1    2019-09-23                   6       317.811667   
47434   1511345        FX0    2019-09-23                   6       294.531667   
47435   1511171        FX0    2019-09-23                   6       245.980000   
47436   1511265        FX0    2019-09-23                   6       526.453333   

       QTD_FATURAS_ULT_6M_FX1  
0                           1  
1                           2  
2                           1  
3                           1  
4                           3  
...                       ...  
47432                       1  
47433                       1  
47434                       2  
47435                       2  
47436                       2  

[47437 rows x 6 columns]